Sources/StructuralExamples/StaticAllKeyPaths.swift

```diff
@@ -0,0 +1,62 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import StructuralCore
+
+/// A protocol that returns key paths for all struct properties:
+protocol StaticAllKeyPaths {
+
+    /// An array of partial key paths for Self. 
+    static var staticAllKeyPaths: [PartialKeyPath<Self>] { get }
+}
+
+extension StaticAllKeyPaths
+where Self: StaticStructural, Self.StructuralRepresentation: _CollectKeyPaths {
+    static var staticAllKeyPaths: [PartialKeyPath<Self>] {
+        var keyPaths: [AnyKeyPath] = []
+        self.staticStructuralRepresentation.collectKeyPaths(into: &keyPaths)
+        return keyPaths as! [PartialKeyPath<Self>]
+    }
+}
+
+protocol _CollectKeyPaths {
+    func collectKeyPaths(into keyPaths: inout [AnyKeyPath])
```

Comment:
What about:

```suggestion
    func collectKeyPaths(into keyPaths: inout [PartialKeyPath<Self>])
```

or:

```suggestion
    associatedtype BaseType
    func collectKeyPaths(into keyPaths: inout [PartialKeyPath<BaseType>])
```
?

Reply:
This complicated implementation of _CollectKeyPath because we don't have `BaseType` for StructuralCons and StructuralEmpty. Adding them just for this use case seems like a massive overkill.

Reply:
Agreed that adding `BaseType` to `StructuralCons` doesn't make sense (although it can likely be conditionally inferred with the "BaseTypeProtocol" I've used in related PRs), and `StructuralEmpty` of course "should" be `Never` which neatly solves that problem.

> I say "should" because I'm not sure how I feel about it yet. While it has some aspect of feeling more regular, I'm not sure it substantially simplifies use of SGP in practice.